packages/tpetra/core/src/Tpetra_CrsSingletonFilter_LinearProblem_def.hpp

```diff
@@ -0,0 +1,919 @@
+    rvs() {
+  // GOT TO HERE!!
+  // ComputeFullSolution();
+}
+
+//==============================================================================
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void CrsSingletonFilter_LinearProblem<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+    Analyze(const Teuchos::RCP<row_matrix_type>& fullMatrix) {
+  const char tfecfFuncName[] = "Analyze: ";
+
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(AnalysisDone_, std::runtime_error,
+                                        "Analyze() already done once.  Cannot do it again.");
+
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(fullMatrix == Teuchos::null,
+                                        std::runtime_error, "Input matrix is Teuchos::null.");
+
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(fullMatrix->getGlobalNumRows() == 0,
+                                        std::runtime_error, "Full matrix has zero dimension.");
+
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(fullMatrix->getGlobalNumEntries() == 0,
+                                        std::runtime_error, "Full matrix has no nonzero terms.");
+
+  FullMatrix_ = fullMatrix;
+
+  // First check for columns with single entries and find columns with singleton rows
+  vector_type_int ColProfiles(FullMatrixColMap());
+  ColProfiles.putScalar(0);
+  vector_type_int ColHasRowWithSingleton(FullMatrixColMap());
+  ColHasRowWithSingleton.putScalar(0);
+
+  // localRowIDofSingletonCol[j] will contain the local row ID associated with the jth column,
+  // if the jth col has a single entry
+  vector_type_LO localRowIDofSingletonCol(FullMatrixColMap());
+  localRowIDofSingletonCol.putScalar(-1);
+
+  // Define MapColoring objects
+  RowMapColors_ = Teuchos::rcp(new vector_type_int(FullMatrixRowMap()));  // Initial colors are all 0
+  ColMapColors_ = Teuchos::rcp(new vector_type_int(FullMatrixColMap()));
+
+  local_ordinal_type localNumRows = FullMatrix()->getLocalNumRows();
+  local_ordinal_type localNumCols = FullMatrix()->getLocalNumCols();
+
+  // Set up for accessing full matrix.  Will do so row-by-row.
+  InitFullMatrixAccess();
+
+  // Scan matrix for singleton rows, build up column profiles
+  size_t NumIndices = 1;
+  // int * localIndices;
+  // nonconst_local_inds_host_view_type localIndices;
+  Teuchos::Array<local_ordinal_type> localIndices;
+  localNumSingletonRows_ = 0;
+
+  auto ColProfilesData              = ColProfiles.getLocalViewHost(Tpetra::Access::ReadWrite);
+  auto localRowIDofSingletonColData = localRowIDofSingletonCol.getLocalViewHost(Tpetra::Access::ReadWrite);
+  auto ColHasRowWithSingletonData   = ColHasRowWithSingleton.getLocalViewHost(Tpetra::Access::ReadWrite);
+  auto RowMapColors_Data            = RowMapColors_->getLocalViewHost(Tpetra::Access::ReadWrite);
+  auto ColMapColors_Data            = ColMapColors_->getLocalViewHost(Tpetra::Access::ReadWrite);
+
+  for (int i = 0; i < localNumRows; i++) {
```

Comment:
This should be a `parallel_reduce`. Something like
```
auto lclMatrix = matrix->getLocalMatrixDevice();
bool abort = false;
Kokkos::parallel_for("some label", Kokkos::RangePolicy<size_t>(0, NumLocalIndeices),
KOKKOS_LAMBDA(const size_t i, bool &needToAbort) {
  auto row = lclMatrix.rowConst(i);
  if (bounds_check_fails) {
    needToAbort = true;
  }
  ......
},  Kokkos::LOr<bool>(abort));
```

Reply:
@iyamazaki ?

Reply:
Outdated.